// node_adcloud/lib/coder/message_tag.js




exports.ADCLOUD_MESSAGE_TAG_APP_CLOSE_CLIENT = 4002
exports.ADCLOUD_MESSAGE_TAG_APP_SCOPE_OP = 4003
exports.ADCLOUD_MESSAGE_TAG_APP_SCOPE_DATA = 4004
exports.ADCLOUD_MESSAGE_TAG_APP_IN = 4005
exports.ADCLOUD_MESSAGE_TAG_APP_MSG = 4006
exports.ADCLOUD_MESSAGE_TAG_APP_OUT = 4007
exports.ADCLOUD_MESSAGE_TAG_APP_OTHER_APP_DATA = 4008
exports.ADCLOUD_MESSAGE_TAG_APP_BROADCAST = 4009

exports.ADCLOUD_MESSAGE_TAG_RPC_REQ = 1001;
exports.ADCLOUD_MESSAGE_TAG_RPC_RSP = 1002;


exports.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST = 5101
exports.ADCLOUD_MESSAGE_TAG_CLIENT_MESSAGE = 5102
exports.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST_BACK = 5103


exports.ADCLOUD_MESSAGE_TAG_CHANNEL_OPEN = 6001
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_OPENBACK = 6002
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_REOPEN = 6003
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_REOPEN_BACK= 6004
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_MESSAGE = 6005
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_MESSAGE_BACK = 6006
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_CLOSE = 6007
exports.ADCLOUD_MESSAGE_TAG_CHANNEL_KEEP_ALIVE = 6008


exports.ADCLOUD_MESSAGE_TAG_PROXY_GET_MASTER  = 7000
exports.ADCLOUD_MESSAGE_TAG_PROXY_GET_MASTER_BACK  = 7001

exports.ADCLOUD_MESSAGE_TAG_PROXY_GET_GATE =  7002
exports.ADCLOUD_MESSAGE_TAG_PROXY_GET_GATE_BACK  = 7003
