// node_adcloud/lib/client_message.js
var coder = require("./coder");
var util = require("util");
exports.MClientReg = MClientReg;
exports.MClientMessage = MClientMessage;

coder.RegistMessage(coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST,MClientReg);
coder.RegistMessage(coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST_BACK,MClientRegBack);
coder.RegistMessage(coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_MESSAGE,MClientMessage);
function MClientReg(appid){
	coder.Message.call(this,coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST);
	this.appid = appid;
}
util.inherits(MClientReg, coder.Message);
MClientReg.prototype.BodyLength = function(){
	var size = Buffer.byteLength(this.appid)  + 1;
	return size;
}

MClientReg.prototype.EncodeBody= function(buf_writer){
	buf_writer.WriteStringAscii(this.appid);
}

MClientReg.prototype.DecodeBody = function(buf_reader){
	this.appid = buf_reader.ReadStringAscii();
}

function MClientRegBack(appid){
	coder.Message.call(this,coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_REGIST_BACK);
	this.code = 0;
}
util.inherits(MClientRegBack, coder.Message);
MClientRegBack.prototype.BodyLength = function(){
	return 1;
}

MClientRegBack.prototype.EncodeBody= function(buf_writer){
	buf_writer.WriteUInt8(this.code);
}

MClientRegBack.prototype.DecodeBody = function(buf_reader){
	this.code = buf_reader.ReadUInt8();
}


function MClientMessage(buf){
	coder.Message.call(this,coder.TAG.ADCLOUD_MESSAGE_TAG_CLIENT_MESSAGE);
	this.buf_ = buf;
}
util.inherits(MClientMessage, coder.Message);
MClientMessage.prototype.BodyLength = function(){
	var size = this.buf_.length;

	return size;
}

MClientMessage.prototype.EncodeBody= function(buf_writer){
	if(this.buf_){
				buf_writer.WriteBuffer(this.buf_);
	}
}

MClientMessage.prototype.DecodeBody = function(buf_reader){
	var buf = buf_reader.GetDataBuffer();
	if(buf.length  == 0){
		return;
	}
	this.buf_ = new Buffer(buf.length);
	buf.copy(this.buf_);
}
